[PATCH] ARM: Ignore memory tags with invalid data

From: Corey Minyard <minyard@acm.org>

The DNS-323 system has several bogus memory entries in the tag table,
and it caused the system to crash at startup.  Ignore tag entries that
are obviously bogus.

Signed-off-by: Corey Minyard <minyard@acm.org>

diff --git a/arch/arm/kernel/setup.c b/arch/arm/kernel/setup.c
index d3941a7..0bb917f 100644
--- a/arch/arm/kernel/setup.c
+++ b/arch/arm/kernel/setup.c
@@ -631,7 +631,12 @@ __tagtable(ATAG_CORE, parse_tag_core);
 
 static int __init parse_tag_mem32(const struct tag *tag)
 {
-	if (meminfo.nr_banks >= NR_BANKS) {
+	/*
+	* Make sure that the memory size is non-zero, page aligned,
+	* and that it doesn't overflow the meminfo table.
+	*/
+	if (meminfo.nr_banks >= NR_BANKS || tag->u.mem.size & ~PAGE_MASK ||
+	   tag->u.mem.size == 0 || tag->u.mem.start & ~PAGE_MASK) {
 		printk(KERN_WARNING
 		       "Ignoring memory bank 0x%08x size %dKB\n",
 			tag->u.mem.start, tag->u.mem.size / 1024);

